src/ic-cdk/src/timer.rs

```diff
@@ -0,0 +1,235 @@
+    );
+    let key = TASKS.with(|tasks| tasks.borrow_mut().insert(Task::Once(Box::new(func))));
+    TIMERS.with(|timers| {
+        timers.borrow_mut().push(Timer {
+            task: key,
+            time: scheduled_time,
+        });
+    });
+    update_ic0_timer();
+    key
+}
+
+/// Sets `func` to be executed every `interval`. Panics if `interval` + [`time()`][crate::api::time] is more than [`u64::MAX`] nanoseconds.
+///
+/// To cancel the interval timer, pass the returned `TimerId` to [`clear_timer`].
+///
+/// Note that timers are not persisted across canister upgrades.
+pub fn set_timer_interval(interval: Duration, func: impl FnMut() + 'static) -> TimerId {
+    let interval_ns = u64::try_from(interval.as_nanos()).expect(
+        "delay out of bounds (must be within `u64::MAX - ic_cdk::api::time()` nanoseconds)",
+    );
+    let scheduled_time = crate::api::time().checked_add(interval_ns).expect(
+        "delay out of bounds (must be within `u64::MAX - ic_cdk::api::time()` nanoseconds)",
+    );
+    let key = TASKS.with(|tasks| {
+        tasks.borrow_mut().insert(Task::Repeated {
+            func: Box::new(func),
+            interval,
+        })
+    });
+    TIMERS.with(|timers| {
+        timers.borrow_mut().push(Timer {
+            task: key,
+            time: scheduled_time,
+        })
+    });
+    update_ic0_timer();
+    key
+}
+
+/// Cancels an existing timer. Does nothing if the timer has already been canceled.
+pub fn clear_timer(id: TimerId) {
+    TASKS.with(|tasks| tasks.borrow_mut().remove(id));
+}
+
+/// Calls `ic0.global_timer_set` with the soonest timer in [`TIMERS`]. This is needed after inserting a timer, and after executing one.
+fn update_ic0_timer() {
+    TIMERS.with(|timers| {
+        let timers = timers.borrow();
+        let soonest_timer = timers.peek().map_or(0, |timer| timer.time);
+        unsafe { ic0::global_timer_set(soonest_timer as i64) };
+    });
+}
+
+#[export_name = "canister_update <ic-cdk internal> timer_executor"]
+extern "C" fn timer_executor() {
+    if crate::api::caller() != crate::api::id() {
+        crate::trap("This function is internal to ic-cdk and should not be called externally.");
+    }
+    let (task_id,) = crate::api::call::arg_data();
+    let task_id = TimerId(KeyData::from_ffi(task_id));
+    // We can't be holding `TASKS` when we call the function, because it may want to schedule more tasks.
+    // Instead, we swap the task out in order to call it, and then either swap it back in, or remove it.
+    let task = TASKS.with(|tasks| {
+        let mut tasks = tasks.borrow_mut();
+        tasks.get_mut(task_id).map(mem::take)
+    });
+    if let Some(mut task) = task {
+        match task {
+            Task::Once(func) => {
+                func();
+                TASKS.with(|tasks| tasks.borrow_mut().remove(task_id));
+            }
+            Task::Repeated { ref mut func, .. } => {
+                func();
```

Comment:
Any concerns about the effects of a panic in `func()`?

Reply:
Yes, well-commented ones 😛 
The inter-canister call is specifically to guard against this possibility, which will catch such a trap. No locks are held across awaits, so such a trap will roll back to a coherent state; no locks are held while the function executes, so resetting the trap point by making a further inter-canister call inside it will still reset it to a coherent state. The code that removes or reschedules the timer is placed outside the inter-canister call, so no matter what happens inside it, the scheduling will behave as intended.